dali/imgcodec/image_decoder_test.cc

```diff
@@ -323,136 +505,313 @@ class ImageDecoderTest : public ::testing::Test {
+    ExpectSuccess(out.res[i]);
+    this->CompareData(out.view[i++], jpeg2000_samples[2]);
+  } else {
+    // we didn't enable jpeg2000 for CPUBackend in this test
+    EXPECT_FALSE(out.res[i++].success);
+    EXPECT_FALSE(out.res[i++].success);
+    EXPECT_FALSE(out.res[i++].success);
+  }
 
   ExpectSuccess(out.res[i]);
-  AssertEqualSatNorm(out.view[i++], jpeg_samples[2].ref);
+  this->CompareData(out.view[i++], jpeg_samples[2]);
+}
+
+TYPED_TEST(ImageDecoderTest_OnlyCPU, DecodeSample_JPEG_SingleDecoder_LibjpegTurbo) {
+#if defined(DALI_USE_JPEG_TURBO)
+  this->FilterDecoder(
+    [](ImageDecoderFactory *factory) {
+      return dynamic_cast<LibJpegTurboDecoderFactory *>(factory) != nullptr;
+    });
+#else
+  GTEST_SKIP();
+#endif
+  this->TestSingleFormatDecodeSample("JPEG");
+}
+
+TYPED_TEST(ImageDecoderTest_OnlyCPU, DecodeSample_JPEG_SingleDecoder_OpenCV) {
+  this->FilterDecoder(
+    [](ImageDecoderFactory *factory) {
+      return dynamic_cast<OpenCVDecoderFactory *>(factory) != nullptr;
+    });
+  this->TestSingleFormatDecodeSample("JPEG");
+}
+
+TYPED_TEST(ImageDecoderTest_OnlyCPU, DecodeSample_TIFF_SingleDecoder_Libtiff) {
+#if LIBTIFF_ENABLED
+  this->FilterDecoder(
+    [](ImageDecoderFactory *factory) {
+      return dynamic_cast<LibTiffDecoderFactory *>(factory) != nullptr;
+    });
+#else
+  GTEST_SKIP();
+#endif
+  this->TestSingleFormatDecodeSample("TIFF");
+}
+
+TYPED_TEST(ImageDecoderTest_OnlyCPU, DecodeSample_TIFF_SingleDecoder_OpenCV) {
+  this->FilterDecoder(
+    [](ImageDecoderFactory *factory) {
+      return dynamic_cast<OpenCVDecoderFactory *>(factory) != nullptr;
+    });
+  this->TestSingleFormatDecodeSample("TIFF");
+}
+
+TYPED_TEST(ImageDecoderTest_OnlyGPU, DecodeSample_JPEG_SingleDecoder_NvJpeg) {
+#if NVJPEG_ENABLED
+  this->FilterDecoder(
+    [](ImageDecoderFactory *factory) {
+      return dynamic_cast<NvJpegDecoderFactory *>(factory) != nullptr;
+    });
+#else
+  GTEST_SKIP();
+#endif
+  this->TestSingleFormatDecodeSample("JPEG");
+}
+
+TYPED_TEST(ImageDecoderTest_OnlyGPU, DecodeSample_JPEG_SingleDecoder_NvJpeg2000) {
+#if NVJPEG2K_ENABLED
+  this->FilterDecoder(
+    [](ImageDecoderFactory *factory) {
+      return dynamic_cast<NvJpeg2000DecoderFactory *>(factory) != nullptr;
+    });
+#else
+  GTEST_SKIP();
+#endif
+  this->TestSingleFormatDecodeSample("JPEG2000");
 }
 
-TYPED_TEST(ImageDecoderTest_CPU, DecodeSample_CorruptedData_JPEG) {
-  this->DisableFallback();
+TYPED_TEST(ImageDecoderTest_CorruptedData, DecodeSample_JPEG) {
+  if (this->IsGPUBackend()) {
+#if NVJPEG_ENABLED
+    this->FilterDecoder([](ImageDecoderFactory *factory) {
+      return dynamic_cast<NvJpegDecoderFactory *>(factory) != nullptr;
+    });
+#else
+    GTEST_SKIP();
+#endif
+  } else {
+#if defined(DALI_USE_JPEG_TURBO)
+    this->FilterDecoder([](ImageDecoderFactory *factory) {
+      return dynamic_cast<LibJpegTurboDecoderFactory *>(factory) != nullptr;
+    });
+#else
+    GTEST_SKIP();
+#endif
+  }
```

Comment:
I think I added so that the decoding actually fails, as the fallback silently reports success (producing an incomplete image). I removed the decoder filter and adjust the expectations.